src/main/java/org/apache/spark/shuffle/ucx/memory/MemoryPool.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2019. ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+package org.apache.spark.shuffle.ucx.memory;
+
+import org.apache.spark.shuffle.UcxShuffleConf;
+import org.apache.spark.unsafe.Platform;
+import org.openucx.jucx.ucp.UcpContext;
+import org.openucx.jucx.ucp.UcpMemory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.nio.ByteBuffer;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+/**
+ * Utility class to reuse and preallocate registered memory to avoid memory allocation
+ * and registration during shuffle phase.
+ */
+public class MemoryPool implements Closeable {
+  private static final Logger logger = LoggerFactory.getLogger(MemoryPool.class);
+
+  @Override
+  public void close() {
+    for (AllocatorStack stack: allocStackMap.values()) {
+      stack.close();
+      logger.info("Stack of size {}. " +
+          "Total requests: {}, total allocations: {}, preAllocations: {}",
+        stack.length, stack.totalRequests.get(), stack.totalAlloc.get(), stack.preAllocs.get());
+    }
+    allocStackMap.clear();
+  }
+
+  private class AllocatorStack implements Closeable {
+    private final AtomicInteger totalRequests = new AtomicInteger(0);
+    private final AtomicInteger totalAlloc = new AtomicInteger(0);
+    private final AtomicInteger preAllocs = new AtomicInteger(0);
+    private final ConcurrentLinkedDeque<RegisteredMemory> stack = new ConcurrentLinkedDeque<>();
+    private final int length;
+
+    private AllocatorStack(int length) {
+      this.length = length;
+    }
+
+    private RegisteredMemory get() {
+      RegisteredMemory result = stack.pollFirst();
+      if (result == null) {
+        ByteBuffer buffer = Platform.allocateDirectBuffer(length);
+        UcpMemory memory = null;
+        if (conf.preregisterMemory()) {
+          memory = context.registerMemory(buffer);
+        }
+        result = new RegisteredMemory(new AtomicInteger(1), memory, buffer);
+        totalAlloc.incrementAndGet();
+      } else {
+        result.getRefCount().incrementAndGet();
+      }
+      totalRequests.incrementAndGet();
+      return result;
+    }
+
+    private void put(RegisteredMemory registeredMemory) {
+      registeredMemory.getRefCount().decrementAndGet();
+      stack.addLast(registeredMemory);
```

Comment:
shouldn't we add to stack only if refcount became 0?

Reply:
but if it's added to the stack any1 can re-use it. wouldn't it create data corruption?

Reply:
No, we put and get from the stack RegisteredMemory class which consists of unique buffer and shared memory region with it's refcount. Application needs only ByteBuffer. Refcount is used only in 1 place actualy: https://github.com/openucx/sparkucx/pull/8/files#diff-379f9437ea127205af26cd258d7ba63fR37 - when we deregister memory.
Here's an example:
Stack:
[----------------------------10Mb---------------------------------] Memory (refcount 0)
[---1Mb--][---1MB--][---1MB--][---1MB--][---1MB--][---1MB--] Buffers

Now you get1 buffer from the stack:
[----------------------------10Mb---------------------------------] Memory (refcount 1)
[---1Mb--][---1MB--][---1MB--][---1MB--][---1MB--]                  Buffers

It means that it's unsafe to deregister 10Mb memory, since there's references to it.

Now you return back buffer to stack:
[----------------------------10Mb---------------------------------] Memory (refcount 0)
[---1Mb--][---1MB--][---1MB--][---1MB--][---1MB--][---1MB--] Buffers

It means that you can now deregister memory and clean a stack. That's only the case with preallocation. With normal alocation memory == buffer.

Reply:
so you mean refcount is for registration and not for the buffer?

Reply:
Correct. When user returns buffer to mempool we assume there's no more references on the user side. In java there's similar stuff to unique_ptr, but this will overcomplicate stuff.